### ISSUE TITLE
GPU Softmax over last dimension of 3D tensor is slow

### ISSUE THREAD

**Issue by nikitakit** [pytorch/pytorch#4893]
Recently I profiled one of my models with `nvprof` and was surprised to find that the softmax layers for an attention mechanism were the most expensive entries in the `cuda_total_time` column.

My original code looked something like this:
```
self.softmax = nn.Softmax(dim=-1)
# ...
# attn is a 3D tensor (batch_size x length x length). Length is in the 30-200 range
attn = self.softmax(attn)
```

After some experimentation, I tried adding some transposes to the code:

```
self.softmax = nn.Softmax(dim=1)
attn = self.softmax(attn.transpose(1, 2)).transpose(1, 2)
```

This increased my overall model speed by around 10%, and now matrix multiplication is at the top of `nvprof` (which is what I would expect).

Could this be considered a performance bug? I wonder if there is some way for the softmax cuda code to have comparable speed regardless of the softmax dimension.

(Sorry I don't have sample code at the moment; my actual code is deeply embedded in the current project I'm doing.)

System info
- OS: Linux
- PyTorch version: 0.3.0
- How you installed PyTorch (conda, pip, source): conda
- Python version: 3.6
- CUDA/cuDNN version: CUDA 9
- GPU models and configuration: K80

**Comment by apaszke** [comment]
Can you give us some example sizes you tried to run on? We need this to do the benchmarks on our side

**Comment by nikitakit** [comment]
I'm applying softmax to batch_size x length x length tensors, where batch size is 8-150 (typically around 90, and always a multiple of 8) and length is 30-200.

Some concrete sizes I sampled from running my code:
```
torch.Size([88, 194, 194])
torch.Size([104, 160, 160])
torch.Size([96, 181, 181])
torch.Size([96, 178, 178])
```

**Comment by apaszke** [comment]
I looked into it and can reproduce the same numbers. Transpose improves the performance 2x for me. I'll look into it, thanks for the report.

**Comment by nikitakit** [comment]
Thanks for looking into this!

**Comment by soumith** [comment]
fixed via #4973
